ui/src/Main.purs

```diff
@@ -57,10 +65,49 @@ update s = case _ of
   SetPage newPage -> pure $ s { currentPage = newPage }
 
   DeviceMsg deviceMsg ->
-    Devices.update s.devices deviceMsg # bimap DeviceMsg (s { devices = _ })
+    Devices.update s.websocket s.devices deviceMsg # bimap DeviceMsg (s { devices = _ })
 
   InitWS ws -> do
-    forks $ \ms -> initializeListeners ws (ms <<< DeviceMsg)
+    forks $ \msgSink -> do
+      let msgSink' = msgSink <<< DeviceMsg
+      el <- liftEffect $ eventListener $ \evt -> do
+        for_ (fromEvent evt) \msgEvt -> do
+          let
+            jsonStr = unsafeFromForeign $ data_ msgEvt
+            jsonBlob = parseJson jsonStr
+            devices = Devices.decodeDevices =<< jsonBlob
+            deviceState = DeviceState.decodeDeviceState =<< jsonBlob
+
+          debug jsonStr
+
+          --
+          -- Probably going to abstract this pattern away.
+          -- What I really want is something that lets me run a bunch
+          -- of Eithers and collect the Left values until I hit a
+          -- Right. That is, I want the short-circuiting behavior of
+          -- `<|>` with something that returns the accumulated state
+          -- of all Left values at the end. And the accumulated state
+          -- should tell me in order what it failed at and finally
+          -- what it succeeded with. Also I think I only want one
+          -- message for parsing failure that summarizes what failed,
+          -- and sends it to debug by default, and warns only if
+          -- nothing was parsed successfully.
+          --
```

Comment:
apparently what I want is https://book.purescript.org/chapter7.html#applicative-validation-1